DOCS-915: Document entry point file for modules and improve modules documentation (#1476)
viamrobotics/docs: docs/extend/modular-resources/_index.md

---

> title: "Integrate Modular Resources into your robot"

Reviewer: ```suggestion
title: "Integrate Modular Resources into your Robot"
```

    - "/program/extend/modular-resources/"
---

The Viam module system allows you to integrate custom {{< glossary_tooltip term_id="resource" text="resources" >}} into any robot running on Viam.

> The {{< glossary_tooltip term_id="rdk" text="RDK">}} [manages](/extend/modular-resources/key-concepts/) these modular resources like the resource types and models that are built-in to the kit.

Reviewer: ```suggestion
The {{< glossary_tooltip term_id="rdk" text="RDK">}} [manages](/extend/modular-resources/key-concepts/) modular resources and built-in resources in the same way.
```
or can we avoid talking about RDK?
```suggestion
`viam-server` [manages](/extend/modular-resources/key-concepts/) modular resources and built-in resources in the same way.
```
(add link on viam-server maybe--and maybe we should add viam-server to the glossary but maybe that's for a different ticket)

Reviewer 2: the latter I think

Modular resources can be:

1. New models of built-in [components](/components/) or [services](/services/) that implement the built-in resource type's API through Viam's [client SDKs](/program/apis/).
2. Brand new types of resources that define their own API in [protocol buffers](https://developers.google.com/protocol-buffers).

## Get Started

To get started adding custom resources to your robot, learn the [key concepts](/extend/modular-resources/key-concepts/) behind Viam's resource APIs that make the module system possible.

> Then, follow one of these [example tutorials](/extend/modular-resources/examples/) to add previously built modules and modular resources to your robot.

Reviewer: But some of these examples show creating new models? Perhaps it would make more sense to point them to the extend/modular-resources/configure/ page instead? And if necessary, that page could have links to specifically relevant examples.

Author: What do you mean? It's the rplidar and odrive, which I think are good simple pages to show the configuration process for pre-built modules. They do offer new models but that's in the module so I don't catch your drift

Reviewer: Right now, the page looks like this: https://docs.viam.com/extend/modular-resources/examples/ It is true that those pre-built modules are contained in the examples folder, but we also have links to tutorials there that create new modules.

I think you should do this:
```suggestion
To add a previously built module to your robot, [configure a modular resource](/extend/modular-resources/configure/].
```

Then, if you want, links to relevant examples can be added to the bottom of the configure page.

Also, somewhat less importantly, this suggestion changes the order of words such that people know what they're doing before clicking the link (versus telling them to click the link before we actually know whether they want to use a prebuilt module or to make a new one).

> If you want to create your own module providing custom resource models or new resource types as modular resources, follow [these instructions](/extend/modular-resources/create/) instead.

Reviewer: ```suggestion
If you want to create your own module providing a new resource model or a new resource type as a modular resource, follow [these instructions](/extend/modular-resources/create/).
```

> After that, follow [these steps](/extend/modular-resources/configure/) to configure the module and modular resource.

Reviewer: ```suggestion
Then, follow [these steps](/extend/modular-resources/configure/) to configure the module and modular resource.
```

    - "/program/extend/modular-resources/"
---

The Viam module system allows you to integrate custom {{< glossary_tooltip term_id="resource" text="resources" >}} into any robot running on Viam.
`viam-server` [manages](/extend/modular-resources/key-concepts/) modular resources and built-in resources in the same way.

Modular resources can be:

1. New models of built-in [components](/components/) or [services](/services/) that implement the built-in resource {{< glossary_tooltip term_id="api-namespace-triplet" text="subtype" >}}'s API through Viam's [client SDKs](/program/apis/).
2. Brand new types of resources that define their own API in [protocol buffers](https://developers.google.com/protocol-buffers).

## Get Started

To get started adding custom resources to your robot, learn the [key concepts](/extend/modular-resources/key-concepts/) behind Viam's resource APIs that make the module system possible.

- To create your own module and modular resources, follow [these instructions](/extend/modular-resources/create/).

> - To download modules and add modular resources that other users have built to your robot, follow one of Viam's [module setup guides](/extend/modular-resources/examples/).

Author: Hmmm IMO it's not essentially the same as configure though-- eventually those module docs that are up on the site will probably mostly be the module descriptions and install instructions + what specific config attributes

From my view this bullet is helpful for people that want to browse options for pre-built modules and are not set on questing the create
